test/unit/spec/VerifyCustomFactor_spec.js

```diff
@@ -0,0 +1,263 @@
+import Router from 'LoginRouter';
+import Beacon from 'helpers/dom/Beacon';
+import VerifyCustomFactorForm from 'helpers/dom/VerifyCustomFactorForm';
+import Util from 'helpers/mocks/Util';
+import Expect from 'helpers/util/Expect';
+import resAllFactors from 'helpers/xhr/MFA_REQUIRED_allFactors';
+import resNoPermissionError from 'helpers/xhr/NO_PERMISSION_error';
+import resChallengeClaimsProvider from 'helpers/xhr/MFA_CHALLENGE_ClaimsProvider';
+import resSuccess from 'helpers/xhr/SUCCESS';
+import $sandbox from 'sandbox';
+import LoginUtil from 'util/Util';
+const SharedUtil = internal.util.Util;
+const itp = Expect.itp;
+
+Expect.describe('VerifyCustomFactor', function() {
+  function createRouter(baseUrl, authClient, successSpy, settings, startRouter) {
+    const router = new Router(
+      _.extend(
+        {
+          el: $sandbox,
+          baseUrl: baseUrl,
+          authClient: authClient,
+          globalSuccessFn: successSpy,
+        },
+        settings
+      )
+    );
+
+    Util.registerRouter(router);
+    Util.mockRouterNavigate(router, startRouter);
+    return router;
+  }
+
+  function expectError(test, code, message, controller, resError) {
+    expect(test.form.hasErrors()).toBe(true);
+    expect(test.form.errorMessage()).toBe(message);
+    expectErrorEvent(test, code, message, controller, resError);
+  }
+
+  function expectErrorEvent(test, code, message, controller, resError) {
+    expect(test.afterErrorHandler).toHaveBeenCalledTimes(1);
+    expect(test.afterErrorHandler.calls.allArgs()[0]).toEqual([
+      {
+        controller: controller,
+      },
+      {
+        name: 'AuthApiError',
+        message: message,
+        statusCode: code,
+        xhr: resError,
+      },
+    ]);
+  }
+
+  function expectSubtitleToBe(test, desiredSubtitle) {
+    expect(test.form.subtitleText()).toBe(desiredSubtitle);
+  }
+
+  async function setup(res, selectedFactorProps, settings, lastFailedChallengeFactorData, languagesResponse, startRouter) {
+    const setNextResponse = Util.mockAjax();
+    const baseUrl = 'https://foo.com';
+    const authClient = createAuthClient({ issuer: baseUrl, transformErrorXHR: LoginUtil.transformErrorXHR });
+    const successSpy = jasmine.createSpy('success');
+    const afterErrorHandler = jasmine.createSpy('afterErrorHandler');
+    const router = createRouter(baseUrl, authClient, successSpy, settings, startRouter);
+
+    router.on('afterError', afterErrorHandler);
+    if (startRouter) {
+      if (typeof startRouter !== 'boolean') {
+        throw new Error('startRouter parameter should be boolean');
+      }
+      await Expect.waitForPrimaryAuth();
+    }
+
+    setNextResponse(res);
+    if (languagesResponse) {
+      setNextResponse(languagesResponse);
+    }
+    router.refreshAuthState('dummy-token');
+    await Expect.waitForMfaVerify();
+
+    router.appState.set('lastFailedChallengeFactorData', lastFailedChallengeFactorData);
```

Comment:
do we need to explicitly set it? If the response is a failure, then this should be set for you (happens in [RouterUtil](https://github.com/okta/okta-signin-widget/blob/master/src/util/RouterUtil.js#L186))

Reply:
I had to set it explicitly.

Reply:
talked offline about this. Unit tests need this explicitly since it is set in RouterUtil, but these tests do not go through that util file.